src/main/java/InfrastructureManager/MatchMaking/ScoreBasedMM.java

```diff
@@ -0,0 +1,228 @@
+            long currentTime = System.currentTimeMillis();
+            long elapsedTime = currentTime - lastConnectedTime;
+
+            //Deducting score: -1 score for every second (1000ms)
+            long deductingScore = deductScorePerSecond * elapsedTime / 1000;
+            logger.info("time since last connected = {} ms, {} score deduct per second", elapsedTime, deductScorePerSecond);
+
+            long nodeHistoryScore = nodeHistoryWithClient - deductingScore;
+
+            if (nodeHistoryScore < 0) {
+                nodeHistoryScore = 0;
+            }
+            logger.info("node history score = {} = {} - {}", nodeHistoryScore, nodeHistoryWithClient, deductingScore);
+            //Update the score in the history info pack with this new score
+            clientHistory.setHistoryScoreForClient(thisClientID, thisNodeID, nodeHistoryScore);
+
+            //Eliminate the one that's not good (small or equal are ruled out, equal is ruled out because it's better to have some sort of buffer rather than 100% utilization
+            if ((nodeResource <= reqResource) || (nodeNetwork <= reqNetwork) || !nodeIsConnected || pingNumber > ACCEPTABLE_PING) {
+                numberOfUnqualified++;
+                //Skip this node if it doesn't have enough resources, network or not connected
+                logger.info("[{}] is skipped due to lacking of some criterion: failed case: {}/{} \n resources ( {} / {} ) --- network ( {} / {} ) --- connected : {} --- ping ( {} / {} )",
+                        thisNode.getId(), numberOfUnqualified, totalNumberOfNode, nodeResource, reqResource, nodeNetwork, reqNetwork, nodeIsConnected, pingNumber, ACCEPTABLE_PING);
+            } else {
+                //Find the one with the most available resources
+                //DONE: change the criteria that we compare to resource + network + ping + history. Not just available resource like this
+
+                //calculate current node score and find best node
+                logger.info("calculating [{}] score: ", thisNodeID);
+                long thisNodeScore = getScore(nodeResource, nodeNetwork, pingNumber, nodeHistoryScore);
+                //TODO: change this to debug
+                logger.debug("best node's score - this node score:  {} - {} ", bestScore, thisNodeScore);
+                if (bestScore < thisNodeScore) {
+                    bestNode = thisNode;
+                    bestScore = thisNodeScore;
+                }
+            }
+
+        }
+
+        //TODO: implement threshold, quality of service
+        //Check whether it's good enough to match, or we just return the rejectedNode
+        if (numberOfUnqualified >= totalNumberOfNode) {
+            logger.warn("""
+                    >>>>>>>>>>>>>>>>>>>>>>>>>    MATCH MAKING FAILED    <<<<<<<<<<<<<<<<<<<<<<<<  
+                    No nodes in system satisfy required parameters (Network, Resource, Distance)
+                    """);
+            return rejectNode;
```

Comment:
Are you sure we want to throw an exception here rather than a rejectNode?   We only throw exception when things are out of our expectation, but no nodes satisfy the client's requirement is normal. returning reject node would be easier to handle in `assign()` function tbh.

Reply:
@zero212 Exceptions are not just stuff that happen out of our expectation. Exceptions are conditions at which the current procedure can't continue. For example if you try to buy something but you don't have enough money, that is an exception.

Returning a rejectNode is a way of handling exceptions as well (C++ way). However, it lacks a lot of features that you can find in the actual Exception class in Java. For example, where should the exception be handled, and the dependencies that result from that.

Actually, I believe you made a mistake here that you didn't realize. I was able to get a client connected to rejectNode, which of course not how the exception should be handled. It is much easier to detect this type of issues if we use a proper exception definition like in Exception Java class.

And by the way, this is not the only instance where an exception should be used, I tried to mark all of the cases where you should have used exceptions.

Reply:
Done in 7adaa17